activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -51,6 +51,15 @@ def to_sql_and_binds(arel_or_sql_string, binds = [], preparable = nil, allow_ret
       end
       private :to_sql_and_binds
 
+      # Compiles Arel in the given QueryIntent if needed
+      def compile_arel_in_intent(intent) # :nodoc:
+        return unless intent.needs_arel_compilation?
+
+        sql, binds, preparable, allow_retry = to_sql_and_binds(intent.arel, intent.binds, intent.prepare, intent.allow_retry)
+        intent.set_compiled_result(raw_sql: sql, binds: binds, prepare: preparable, allow_retry: allow_retry)
+      end
+      private :compile_arel_in_intent
```

Comment:
I believe this logic should move inside `QueryAttempt`, e.g. `intent.compile(self)`. It would be better encapsulation.